cmd/gpu-kubelet-plugin/driver.go

```diff
@@ -79,68 +80,71 @@ func (d *driver) Shutdown() error {
+	results := make(map[types.UID]kubeletplugin.PrepareResult)
 
-	for _, claim := range req.Claims {
-		preparedResources.Claims[claim.UID] = d.nodePrepareResource(ctx, claim)
+	for _, claim := range claims {
+		results[claim.UID] = d.nodePrepareResource(ctx, claim)
 	}
 
-	return preparedResources, nil
+	return results, nil
 }
 
-func (d *driver) NodeUnprepareResources(ctx context.Context, req *drapbv1.NodeUnprepareResourcesRequest) (*drapbv1.NodeUnprepareResourcesResponse, error) {
-	klog.Infof("NodeUnprepareResource is called: number of claims: %d", len(req.Claims))
-	unpreparedResources := &drapbv1.NodeUnprepareResourcesResponse{Claims: map[string]*drapbv1.NodeUnprepareResourceResponse{}}
+func (d *driver) UnprepareResourceClaims(ctx context.Context, claims []kubeletplugin.NamespacedObject) (map[types.UID]error, error) {
+	klog.Infof("UnprepareResourceClaims called with %d claim(s)", len(claims))
 
-	for _, claim := range req.Claims {
-		unpreparedResources.Claims[claim.UID] = d.nodeUnprepareResource(ctx, claim)
+	results := make(map[types.UID]error)
+
+	for _, claim := range claims {
+		results[claim.UID] = d.nodeUnprepareResource(ctx, claim)
 	}
 
-	return unpreparedResources, nil
+	return results, nil
 }
 
-func (d *driver) nodePrepareResource(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodePrepareResourceResponse {
+func (d *driver) nodePrepareResource(ctx context.Context, claim *resourceapi.ResourceClaim) kubeletplugin.PrepareResult {
 	d.Lock()
 	defer d.Unlock()
 
-	resourceClaim, err := d.client.ResourceV1beta1().ResourceClaims(claim.Namespace).Get(
-		ctx,
-		claim.Name,
-		metav1.GetOptions{})
+	devs, err := d.state.Prepare(ctx, claim)
+
 	if err != nil {
-		return &drapbv1.NodePrepareResourceResponse{
-			Error: fmt.Sprintf("failed to fetch ResourceClaim %s in namespace %s", claim.Name, claim.Namespace),
+		return kubeletplugin.PrepareResult{
+			Err: fmt.Errorf("error preparing devices for claim %v: %v", claim.UID, err),
 		}
 	}
 
-	prepared, err := d.state.Prepare(ctx, resourceClaim)
-	if err != nil {
-		return &drapbv1.NodePrepareResourceResponse{
-			Error: fmt.Sprintf("error preparing devices for claim %v: %v", claim.UID, err),
+	var prepDevs []kubeletplugin.Device
+	for _, d := range devs {
+		device := kubeletplugin.Device{
+			// use stripSubrequestNames() helper?
+			Requests:     d.RequestNames,
+			PoolName:     d.PoolName,
+			DeviceName:   d.DeviceName,
+			CDIDeviceIDs: d.CDIDeviceIDs,
 		}
+		prepDevs = append(prepDevs, device)
 	}
 
-	klog.Infof("Returning newly prepared devices for claim '%v': %v", claim.UID, prepared)
-	return &drapbv1.NodePrepareResourceResponse{Devices: prepared}
+	klog.Infof("Returning newly prepared devices for claim '%v': %v", claim.UID, prepDevs)
+	return kubeletplugin.PrepareResult{Devices: prepDevs}
 }
 
-func (d *driver) nodeUnprepareResource(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodeUnprepareResourceResponse {
+// Is this never retryable?
```

Comment:
CC @klueska here I noticed that in case of CD we plan for retryable unrepare errors, whereas in case of GPU we probably treat each error as permanent. I think I can already see why that is, but maybe we can talk about that a bit.

Reply:
I was just going to comment on this as a particular source of assymmetry between our two drivers. Does it make sense to update the implementation of the GPU driver to match that of the CD driver, with errors always being marked as non-retryable. This can be considered out of scope of this PR though.

Reply:
> match that of the CD driver, with errors always being marked as non-retryable

Hm! You made me question my understanding of the error handling idioms at play here. Good.

I think it's the other way round: most errors in the CD plugin are implicitly marked as retryable (by virtue of _not_ being created as  `permanentError()`.

I found only one code path generating an error explicitly marked with `permanentError()`: https://github.com/NVIDIA/k8s-dra-driver-gpu/blob/d17cfcf0b84cf12d87c3377cd8b811ad0a91ba72/cmd/compute-domain-kubelet-plugin/device_state.go#L335

Reply:
I think we should make them symmetrical (possibly with more permanent errors in this driver). My focus was on the computeDomain driver for the time being though, because this driver will not be enabled by default (nor supported) in our initial release.